File: matlab/CalcAngleFromStress.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Copyright 2026 Robot Squirrel Productions.
%
%          (\_   _/)
%          ( o   o )
%          (  ^_^  )
%             \_/_
%
% ABOUT
% Given shaft section properties and applied torque, this function returns
% the total angle of twist in radians.
% 
% The companion Mathematica worksheet ("GeometricProperties.nb") has
% pictures and defintions of the terms used in this function. This code
% follows the nomenclature defined in 2529RS0011
%
% EXAMPLE
% d_phi = CalcAngleFromStress(d_ro, d_L, d_taumax, d_G);
%
% INPUT
% d_ro (REQUIRED, POSITIVE SCALAR) - Outer radius of circular section,
%   engineering units of m or in.
% d_L (REQUIRED, NON-NEGATIVE SCALAR) - Shaft section length, engineering
%   units of m or in.
% d_taumax (REQUIRED, NUMERIC) - Torsional shear stress. If a vector then
%   the output will also be vector.
% d_G (REQUIRED, POSITIVE SCALAR) - Shear modulus of elasticity, in
%   engineering units of Pascal or psi. This function does not convert
%   units so it must match length and torque units.
%
% OUPUT
% d_phi (SCALAR) - Angle of twist, radians 
%
% B. Howard
% 20 Aug 2024
%
function d_phi = CalcAngleFromStress(d_ro, d_L, d_taumax, d_G, varargin)

% Instantiate the parser
p = inputParser;

% Validation rules
validScalarPosNum = @(x) assert(isnumeric(x) && isscalar(x) && (x > 0),...
    'Must be a positive scalar numeric quantity');

% Define the inputs
addRequired(p, 'd_ro', validScalarPosNum)
addRequired(p, 'd_L', validScalarPosNum)
addRequired(p, 'd_taumax', @isnumeric)
addRequired(p, 'd_G', validScalarPosNum)

% Parse the inputs
parse(p, d_ro, d_L, d_taumax, d_G, varargin{:});

% Bring the inputs into the function
d_ro = p.Results.d_ro;
d_L = p.Results.d_L;
d_taumax = p.Results.d_taumax(:);
d_G = p.Results.d_G;

% Perform the calculation
d_phi = (d_L .* d_taumax  ) ./ ( d_G .* d_ro );

return

File: python/vibration_analysis/torsional/calc_angle_from_stress.py
"""Angle-of-twist utilities from torsional shear stress inputs.

This module mirrors the MATLAB ``CalcAngleFromStress`` behavior.
"""

from __future__ import annotations

from typing import Any

import numpy as np


def calc_angle_from_stress(
    d_ro: float,
    d_L: float,
    d_taumax: Any,
    d_G: float,
) -> np.ndarray:
    """Return shaft twist angle in radians from shear stress.

    The implemented relationship is:
        phi = (L * tau_max) / (G * r_o)

    Args:
        d_ro: Outer radius of circular shaft section (positive scalar).
        d_L: Shaft section length (positive scalar).
        d_taumax: Torsional shear stress, scalar or array-like.
        d_G: Shear modulus of elasticity (positive scalar).

    Returns:
        A column-shaped ``numpy.ndarray`` of twist angle(s) in radians.
    """

    for value, name in ((d_ro, "d_ro"), (d_L, "d_L"), (d_G, "d_G")):
        if not np.isscalar(value) or not np.isfinite(value) or value <= 0.0:
            raise ValueError(f"{name} must be a positive scalar numeric quantity")

    tau_vector = np.asarray(d_taumax, dtype=float).reshape(-1, 1)
    if tau_vector.size == 0 or np.any(~np.isfinite(tau_vector)):
        raise ValueError("d_taumax must contain finite numeric value(s)")

    return (d_L * tau_vector) / (d_G * d_ro)
